File: day1/day1.rb
def calculate_max_calories(filename, number_elves = 1)
  filename = File.dirname(__FILE__) + "/" + filename
  # create an iterator to avoid loading the whole file into memory
  # but also so we can check for the end of the file as we loop
  lines = File.foreach(filename)

  totals = []
  sum = 0

  lines.each_with_index do |line, idx|
    sum += line.to_i
    if /^([^0-9]*)$/.match?(line)
      totals << sum
      sum = 0
    elsif idx == lines.count - 1
      totals << sum
    end
  end

  totals.sort!.reverse!
  totals.first(number_elves).sum
end

def test(calculation, expected)
  puts (calculation == expected) ? "PASS" : "FAIL"
  puts calculation
  puts ""
end

# Tests
test(calculate_max_calories("calories_test.txt"), 24000)
test(calculate_max_calories("calories_test2.txt"), 24000)
test(calculate_max_calories("calories.txt"), 71502)
test(calculate_max_calories("calories_test.txt", 3), 45000)
test(calculate_max_calories("calories_test2.txt", 3), 45000)

# Results
puts calculate_max_calories("calories.txt", 3)

File: day1/day1.js
const fs = require('fs').promises;

const maxCalories = async (filename, numElves = 1) => {
    let currentSum = 0;
    const sums = [];

    const data = await fs.readFile(filename, 'utf8');
    const lines = data.split('\n');

    for (const line of lines) {
        if (line === '' && currentSum !== 0) {
            sums.push(currentSum);
            currentSum = 0;
        } else {
            currentSum += Number(line);
        }
    }
    sums.push(currentSum);
    sums.sort((a, b) => { return a - b });

    let retValue = 0;
    for (i = 0; i < numElves; i++) { retValue += sums.pop(); }
    return retValue;
}

const test = (filename, elves, expectedAnswer) => {
    maxCalories(filename, elves).then(answer => {
        if (answer === expectedAnswer) {
            console.log("Success: " + expectedAnswer);
        } else {
            console.log("=== nope! Got: " + answer + ' but expected: ' + expectedAnswer);
        }
    }).catch(err => {
        console.error(err);
    });
}

test('calories_test.txt', 1, 24000);
test('calories_test2.txt', 1, 24000);
test("calories.txt", 1, 71502);
test("calories_test.txt", 3, 45000)
test("calories_test2.txt", 3, 45000)
test("calories.txt", 3)
